Added editing cycle history

IraSoro opened the pull request:
Closed #87 

I added an edit that is in the menu.

Editing is done by adding or deleting marked days in the calendar. You can edit only up to the possible finish day of the current cycle.

I also added a native back button for it. But it only goes to the ```home tab```. But I think it will be enough for this version.

Screenshots here:
![image](https://github.com/IraSoro/peri/assets/52165881/a87297e3-17e9-4c78-80aa-a26ecdf4d586)
![image](https://github.com/IraSoro/peri/assets/52165881/e2973a45-4ea2-4cd9-b1fb-ec66c77b1008)

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+  const nowDate = new Date();
+  nowDate.setHours(0, 0, 0, 0);
+
+  const isActiveDates = (dateString: string) => {
```
More looks like the calculation logic, maybe it's may be better to move this function in the CalculationLogic file and cover with tests?

IraSoro replied:
Did  it

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+          fill="solid"
+          onClick={() => {
+            if (datetimeRef.current?.value) {
+              const periodDays = [datetimeRef.current.value].flat().sort();
```
You can create `periodDays` const before `return` and use the same constant here and above in the `IonDateTime`'s `value` prop

Now it's looks strange that you calling `periodDays()` function above and here use `string` value of the same array and also `flat().sort()` it

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+            if (datetimeRef.current?.value) {
+              const periodDays = [datetimeRef.current.value].flat().sort();
+              const millisecondsInDay = 24 * 60 * 60 * 1000;
+              periodDays.sort();
```
You already called `sort()` above

IraSoro replied:
Forgot to delete

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+          fill="solid"
+          onClick={() => {
+            if (datetimeRef.current?.value) {
+              const periodDays = [datetimeRef.current.value].flat().sort();
```
Too complicated logic which, in my opinion, should be in the `CalculationLogic` and covered with tests

IraSoro replied:
Did it

imblowfish commented on src/modals/Menu.tsx:
```diff
@@ -133,27 +134,40 @@ const Exporter = () => {
-        slot="start"
-        icon={createOutline}
+    <>
+      <EditModal
```
Just a my opinion, but you don't need to create separated file with `EditModal` component it you'll shorten the code and separate the logic between `СalculationLogic` and the component

IraSoro replied:
I think this modal should be left in a separate file. It has a lot of calculations, large code and has its own design. 
Even after moving logic calculations.

imblowfish replied:
We discussed offline that you'll better to move all calc logic in the `CalculationLogic` file, making `EditModal`  short enough to leave it inside `Menu.tsx`

IraSoro replied:
Yes

IraSoro commented on src/data/Storage.ts:
```diff
@@ -66,7 +66,7 @@ export const storage = {
 // _menstrualPhase(6).catch((err) => console.error(err));
 // _follicularPhase(6).catch((err) => console.error(err));
-// _lutealPhase(6).catch((err) => console.error(err));
+_lutealPhase(6).catch((err) => console.error(err));
```
Yes

IraSoro commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+      startOfCycle.setHours(0, 0, 0, 0);
+      value.push(format(startOfCycle, "yyyy-MM-dd"));
+
+      for (let i = 1; i < cycle.periodLength; i++) {
```
Fixed

IraSoro commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+      value.push(format(startOfCycle, "yyyy-MM-dd"));
+
+      for (let i = 1; i < cycle.periodLength; i++) {
+        startOfCycle.setDate(startOfCycle.getDate() + 1);
```
No. Here, the ```startOfCycle``` increases by 1 every iteration. 
But I corrected this function a little.

IraSoro commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+        <IonItem
+          color="basic"
+          lines="none"
+        ></IonItem>
```
I did this so that the empty space before and after the calendar is equal. But maybe that's not the best way.

IraSoro commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+        <IonItem
+          color="basic"
+          lines="none"
+        ></IonItem>
```
Wrote below

IraSoro commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,152 @@
+          onClick={() => {
+            if (datetimeRef.current?.value) {
+              const periodDays = [datetimeRef.current.value].flat().sort();
+              const millisecondsInDay = 24 * 60 * 60 * 1000;
```
Yes, fixed.

IraSoro commented:
I moved the calculations in the ```CalculationsLogic```. Also added tests for this function.

Also I added the following:
if the user marks cycles like this
![image](https://github.com/IraSoro/peri/assets/52165881/83969a30-9d98-4114-8534-40d68a65deef)

then it counts as one cycle. After calculations and writing to the storage, it will look like this. 
![image](https://github.com/IraSoro/peri/assets/52165881/f5043b48-856b-411d-a02b-4f0d44bc4449)

Let's leave it that way for this version.

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,129 @@
+
+    return (
+      date.getTime() < lastCycleFinish.getTime() ||
+      date.getTime() <= nowDate.getTime()
```
Looks like this expression always will be `true`because any date will be lesser than now date

IraSoro replied:
No, because it is a condition for all calendar dates

imblowfish replied:
Understand, then could you please rename `lastCycleFinish`? Because now this variable looks like last cycle until today, but you mean that this cycle the end of the next expected cycle, which is calculated for the calendar

IraSoro replied:
Renamed

imblowfish commented on src/modals/EditModal.tsx:
```diff
@@ -0,0 +1,129 @@
+      startOfCycle.setHours(0, 0, 0, 0);
+
+      for (let i = 0; i < cycle.periodLength; i++) {
+        const newDate = new Date(startOfCycle);
```
`newDate` is always the same on every loop step, you can move this cojst in the outer loop

IraSoro replied:
No, because in each iteration ```newDate``` is changed and written to the array.
See lines below.

imblowfish replied:
Missed this, agree

imblowfish commented on src/tests/CalculationLogic.test.ts:
```diff
@@ -351,3 +352,67 @@ describe("getPhase", () => {
+    jest.spyOn(i18n, "t").mockImplementation((key) => key);
+
+    const periodDays = [
+      "2023-08-04",
```
Looks like easy to automate

IraSoro replied:
For tests it doesn t matter.

imblowfish replied:
Ok, I thought those dates would be a side effect for the whole test
